feat: implement a Keptn metrics provider

bacherfl opened the pull request:
Closes #1627

This PR adds the implementation of a `keptn` metrics provider

The main difference to other providers in Flagger is that this one is using a Kubernetes client to interact with Keptn resources (`KeptnMetrics` and `Analyses`). Therefore, in addition to the code in `keptn.go`, the `ClusterRole` needed to be adapted to be able to access the required Keptn resources.
The provider supports the use of the following resources:

- `KeptnMetric`: This resource represents the current value of a certain metric. The metric is retrieved by Keptn from one of the following data sources: `prometheus`, `thanos`, `cortex`, `dynatrace`, `datadog`.
- `Analysis`: This Keptn resource can be used to apply a more complex grading logic over a set of different metrics. More information about this concept can be found in the [Keptn Docs](https://keptn.sh/stable/docs/guides/slo/)

stefanprodan commented on docs/gitbook/usage/metrics.md:
```diff
@@ -668,3 +668,49 @@ Reference the template in the canary analysis:
+spec:
+  provider:
+    type: keptn
+  query: |
+    keptnmetric/my-namespace/response-time
```
```suggestion
  query: "keptnmetric/my-namespace/response-time"
```

I would make this a one-liner to suggest you can't have more than one object reference in there. I would also add an example for a real Istio metric like Istio 404 rate https://docs.flagger.app/usage/metrics#prometheus

stefanprodan commented on pkg/controller/controller.go:
```diff
@@ -18,6 +18,7 @@ package controller
 
 import (
 	"fmt"
+	"k8s.io/client-go/rest"
```
Please sort the imports into groups, run `make fmt`

stefanprodan commented on pkg/metrics/providers/keptn.go:
```diff
@@ -0,0 +1,239 @@
+	ctx, cancel := context.WithTimeout(context.Background(), k.analysisTimeout)
+	defer cancel()
+
+	createdAnalysis, err := k.client.
+		Resource(obj.GroupVersionResource).
+		Namespace(obj.Namespace).
+		Create(ctx, analysis, v1.CreateOptions{})
```
I think we should delete the generated objects when we are done reading the result so we don't fill etcd.

aryan9600 commented on pkg/metrics/providers/keptn.go:
```diff
@@ -0,0 +1,252 @@
+
+	// delete the created analysis at the end of the function
+	defer func() {
+		_ = k.client.
```
instead of ignoring the error, we should log it.

aryan9600 commented on pkg/metrics/providers/keptn.go:
```diff
@@ -0,0 +1,252 @@
+	}
+	return &KeptnProvider{
+		client:          client,
+		analysisTimeout: 10 * time.Second,
```
has this been chosen randomly? how does this play with the Canary analysis's interval?

bacherfl replied:
This is the maximum time the provider will wait for the result of a Keptn Analysis to be available - I added this to prevent the provider from blocking the flagger controller indefinitely in case the created Analysis cannot be completed. 
I chose this value as this should be sufficient in most cases for the created Keptn Analysis to be reconciled by the Keptn Analysis controller, however I can also adjust this If you'd like, or if there are some guidelines on how long a provider might take to retrieve a value.

aryan9600 reviewed (commented):
thanks @bacherfl for this! i did a quick first pass, will take a second look soon.

bacherfl commented:
> thanks @bacherfl for this! i did a quick first pass, will take a second look soon.

Thanks for the review @aryan9600! I just addressed your comments and adapted the code accordingly - If there are some further questions feel free to reach out :)

aryan9600 reviewed (commented):
took a second pass, its looking good, just had some minor nits.

can the docs be improved a bit more? a few suggestions:

* the `query` field in `MetricTemplate` example mentions the timeframe and arguments as well
* make it clearer that resource name is actually that of the `AnalysisDefinition` and not the to be created `Analysis`.
* an example around `Analysis`, which shows a sample `AnalysisDefinition` and what the `Analysis` object generated by Flagger would look like

stefanprodan commented:
@aryan9600 could you please push your suggestions on top of this PR. I see that @bacherfl has enabled us to push commit.

bacherfl commented:
Hi @stefanprodan @aryan9600 sorry for the late reply - I had a busy schedule recently. I'm not sure if i can get to incorporate the suggested changes this week, but I may have some time to do so next week - otherwise, feel free to add directly to the PR, as @stefanprodan suggested

aryan9600 commented:
PR updated, the docs could still use a bit more info around `Anlysis`/`AnlysisDefinition`, but we can tackle that in another PR @bacherfl

stefanprodan commented:
@aryan9600 please sync the CRDs, the helm chart has the old one.

stefanprodan reviewed (approved):
LGTM 

Thanks @bacherfl and @aryan9600
